### PR TITLE
add Basic authentication scheme, a Role entity and set access to the endpoints based on the user's role

### PR DESCRIPTION
**Pull request by AntonKurylo**

### CODE REVIEW

File: src/main/java/mate/academy/controller/AuthenticationController.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
 public class AuthenticationController {
     private final UserService userService;
 
-    @PostMapping("/registration")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/controller/AuthenticationController.java]: revert

File: src/main/java/mate/academy/dto/user/UserRegistrationRequestDto.java [hunk 2]
```diff
@@ -21,7 +22,8 @@ public record UserRegistrationRequestDto(
         String firstName,
         @NotBlank
         String lastName,
-        String shippingAddress
+        String shippingAddress,
+        Set<Long> roles
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/user/UserRegistrationRequestDto.java]: remove

File: src/main/java/mate/academy/dto/user/UserResponseDto.java [hunk 3]
```diff
@@ -1,10 +1,16 @@
...
+
+@Getter
+@Setter
+public class UserResponseDto {
+    private Long id;
+    private String email;
+    private String firstName;
+    private String lastName;
+    private String shippingAddress;
+    private Set<Long> roles;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/user/UserResponseDto.java]:
```suggestion
```

File: src/main/java/mate/academy/mapper/UserMapper.java [hunk 4]
```diff
@@ -1,14 +1,38 @@
...
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
 
 @Mapper(config = MapperConfig.class)
 public interface UserMapper {
+    @Mapping(target = "roles", ignore = true)
     UserResponseDto toDto(User user);
 
+    @AfterMapping
+    default void setSkillIds(@MappingTarget UserResponseDto userResponseDto, User user) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/mapper/UserMapper.java]: remove

File: src/main/java/mate/academy/mapper/UserMapper.java [hunk 5]
```diff
@@ -1,14 +1,38 @@
...
+                        .map(Role::getId)
+                                .collect(Collectors.toSet());
+        userResponseDto.setRoles(roleIds);
+    }
+
+    @Mapping(target = "roles", ignore = true)
     User toEntity(UserRegistrationRequestDto requestDto);
+
+    @AfterMapping
+    default void setSkills(@MappingTarget User user, UserRegistrationRequestDto requestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/mapper/UserMapper.java]: remove

File: src/main/java/mate/academy/mapper/UserMapper.java [hunk 6]
```diff
@@ -1,14 +1,38 @@
...
 
+    @AfterMapping
+    default void setSkillIds(@MappingTarget UserResponseDto userResponseDto, User user) {
+        Set<Long> roleIds = user.getRoles().stream()
+                        .map(Role::getId)
+                                .collect(Collectors.toSet());
+        userResponseDto.setRoles(roleIds);
+    }
+
+    @Mapping(target = "roles", ignore = true)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/mapper/UserMapper.java]:
```suggestion
```

File: src/main/java/mate/academy/mapper/UserMapper.java [hunk 7]
```diff
@@ -1,14 +1,38 @@
...
+import mate.academy.model.Role;
 import mate.academy.model.User;
+import org.mapstruct.AfterMapping;
 import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
 
 @Mapper(config = MapperConfig.class)
 public interface UserMapper {
+    @Mapping(target = "roles", ignore = true)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/mapper/UserMapper.java]:
```suggestion
```

File: src/main/java/mate/academy/model/Role.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+@Setter
+public class Role implements GrantedAuthority {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false, unique = true, columnDefinition = "varchar")
+    @Enumerated(EnumType.STRING)
+    private RoleName role;
+
+    public Role(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/model/Role.java]:
```suggestion
```

File: src/main/java/mate/academy/model/User.java [hunk 9]
```diff
@@ -5,15 +5,23 @@
...
 import lombok.Setter;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
 
 @Entity
 @Table(name = "users")
 @Getter
 @Setter
-public class User {
+public class User implements UserDetails {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/model/User.java:28]: Think about how we should add columns like is_deleted. Should we add them directly to existing scripts or there is another better way?

File: src/main/java/mate/academy/security/CustomUserDetailsService.java [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+@Component
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        return userRepository.findByEmail(email).orElseThrow(() ->
+                new UsernameNotFoundException("Can't find user by email"));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/security/CustomUserDetailsService.java]:
```suggestion
                new EntityNotFoundException("Can't find user by email: " + email));
```

File: src/main/java/mate/academy/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -22,8 +25,9 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
             throw new RegistrationException(
                     "The user with the email is already registered: " + requestDto.email());
         }
+        User user = userMapper.toEntity(requestDto);
+        user.setPassword(passwordEncoder.encode(user.getPassword()));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/UserServiceImpl.java:34]: user.setRoles(Set.of(..... 
find role by name USER and set it

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+      author: mateacademy
+      changes:
+        - createTable:
+            tableName: users_roles
+            columns:
+              - column:
+                  name: user_id
+                  type: bigint
+                  constraints:
+                    nullable: false
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml]:
```suggestion
                    foreignKeyName: fk_users_roles_user
                    referencedTableName: users
                    referencedColumnNames:  id
                    nullable:  false
```

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+                    primaryKey: true
+                    primaryKeyName: pk_user_role
+                    foreignKeyName: fk_users_roles_user
+                    references: users(id)
+              - column:
+                  name: role_id
+                  type: bigint
+                  constraints:
+                    nullable: false
+                    primaryKey: true
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml]: the same

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/mate/academy/model/Role.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
+@Table(name = "roles")
+@NoArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/model/Role.java]:
```suggestion
```

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+      changes:
+        - createTable:
+            tableName: users_roles
+            columns:
+              - column:
+                  name: user_id
+                  type: bigint
+                  constraints:
+                    primaryKey: true
+                    primaryKeyName: pk_user_role
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml]: remove
```
primaryKey: true
primaryKeyName: pk_user_role
```

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+                    foreignKeyName: fk_users_roles_user
+                    referencedTableName: users
+                    referencedColumnNames: id
+                    nullable: false
+              - column:
+                  name: role_id
+                  type: bigint
+                  constraints:
+                    primaryKey: true
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml]: remove 
```
primaryKey: true
primaryKeyName: pk_user_role
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see pr

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml (hunk 15 above)
Comment (AntonKurylo) [reply src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml]: The new version of creating the users_roles table creates a unique composite key for each table record. It seems to be better, if not - correct me

Comment (Elena-Bruyako) [reply src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml]: Whar do you mean the new version of creating the users_roles table? 
For my opinion, composite key is used for a little different approach

https://www.baeldung.com/jpa-many-to-many

**Review by Elena-Bruyako — APPROVED** [review]
Answered your question. Good job

**Pull request closed without merging**